### PR TITLE
GoReleaser | added GitHub release workflows.

### PR DESCRIPTION
**Pull request by RutvikS-crest**
## Description

- This workflow will release a new Terraform Provider version to Hashicorp registry when a new tag in vx.x.x format is pushed to the repository.
- added `GPG_PRIVATE_KEY` and `PASSPHRASE` repository secrets with dummy values. we need to create and update respective secrets with a valid [Signing key](https://developer.hashicorp.com/terraform/registry/providers/publishing#preparing-and-adding-a-signing-key).

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [x] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#
+# This uses an action (crazy-max/ghaction-import-gpg) that assumes you set your
+# private key in the `GPG_PRIVATE_KEY` secret and passphrase in the `PASSPHRASE`
+# secret. If you would rather own your own GPG handling, please fork this action
+# or use an alternative one for key handling.
+#
+# You will need to pass the `--batch` flag to `gpg` in your signing step 
+# in `goreleaser` to indicate this is being used in a non-interactive mode.
+#
+name: release
```
Comment (RutvikS-crest) [inline .github/workflows/release.yml:12]: This workflow will release a new Terraform Provider version to Hashicorp registry when a new tag in vx.x.x format is pushed to the repository.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      -
+        name: Run GoReleaser
+        uses: goreleaser/goreleaser-action@f82d6c1c344bcacabba2c841718984797f664a6b # v4.2.0
+        with:
+          version: latest
+          args: release --rm-dist
+        env:
+          GPG_FINGERPRINT: ${{ steps.import_gpg.outputs.fingerprint }}
+          # GitHub sets this automatically
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (cisco-service) [inline .github/workflows/release.yml:51]: [NIT] missing newline

Comment (lelia) [inline .github/workflows/release.yml:51]: [NIT] missing newline

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        with:
+          go-version-file: 'go.mod'
+          cache: true
+      -
+        name: Import GPG key
+        uses: crazy-max/ghaction-import-gpg@111c56156bcc6918c056dbef52164cfa583dc549 # v5.2.0
+        id: import_gpg
+        with:
+          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+          passphrase: ${{ secrets.PASSPHRASE }}
```
Comment (lelia) [inline .github/workflows/release.yml:41]: @RutvikS-crest I've added the GPG key and passphrase as GitHub secrets, matching the above values. Can you confirm whether this workflow requires any further modification or testing, now that the keys are in place?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      - 'README.md'
+  push:
+    paths-ignore:
+      - 'README.md'
+  # For systems with an upstream API that could drift unexpectedly (like most SaaS systems, etc.),
+  # we recommend testing at a regular interval not necessarily tied to code changes. This will 
+  # ensure you are alerted to something breaking due to an API change, even if the code did not
+  # change.
+  # schedule:
+  #   - cron: '0 13 * * *'
```
Comment (lelia) [inline .github/workflows/test.yml:17]: Should we remove this?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        
+    - name: TF acceptance tests
+      timeout-minutes: 10
+      env:
+        TF_ACC: "1"
+
+        # Set whatever additional acceptance test env vars here. You can
+        # optionally use data from your repository secrets using the
+        # following syntax:
+        # SOME_VAR: ${{ secrets.SOME_VAR }}
```
Comment (lelia) [inline .github/workflows/test.yml:105]: Do we need this?

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+      run: |
+        go test -v -cover ./internal/provider/
```
Comment (lelia) [inline .github/workflows/test.yml:108]: [NIT] missing newline

**Review by lelia — CHANGES_REQUESTED** [review]
@RutvikS-crest thanks for putting this together! Some minor NITs and questions about further verification now that the repository has access to the signing keys.

File: .github/workflows/test.yml (hunk 4 above)
Comment (RutvikS-crest) [reply .github/workflows/test.yml:17]: Yep, Removed it

File: .github/workflows/release.yml (hunk 3 above)
Comment (RutvikS-crest) [reply .github/workflows/release.yml:41]: That's it from repo side, I am assuming the key is also registered with Cisco/DevNet space in Hashicorp registry.

Comment (rquedas) [reply .github/workflows/release.yml:41]: @lelia , can you confirm Rutvik's assumption is correct?

Comment (lelia) [reply .github/workflows/release.yml:41]: @rquedas @RutvikS-crest it should now be registered both with Hashicorp and GitHub, yes. There is a `cisco-service` account (shared between CiscoDevNet and a few other Cisco open source orgs) which we have used to upload the new keys to.

**Review by lelia — CHANGES_REQUESTED** [review]
@RutvikS-crest these failing tests may point to a potential problem: https://github.com/cisco-open/terraform-provider-appd/actions/runs/5505021725/jobs/10031873219?pr=35

Looks like it may still be referencing the original urlpath for the Go Client:

```bash
Run go mod download
go: github.com/cisco-open/appd-cloud-go-client@v1.0.2-0.202212280[4](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5505021725/jobs/10031873219?pr=35#step:4:5)3829-8b9388d7[5](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5505021725/jobs/10031873219?pr=35#step:4:6)433: parsing go.mod:
	module declares its path as: github.com/aniketk-crest/appdynamicscloud-go-client
	        but was required as: github.com/cisco-open/appd-cloud-go-client
Error: Process completed with exit code 1.
```

FYI, you should have maintainer access to https://github.com/cisco-open/appd-cloud-go-client in case any further action is needed on the clientside to fix.

File: .github/workflows/release.yml (hunk 3 above)
Comment (lelia) [reply .github/workflows/release.yml:41]: @RutvikS-crest FYI that I'm going to remove the repository-level placeholders for `GPG_PRIVATE_KEY` and `PASSPHRASE`, as they are currently superseding the organization-level secrets that I've added.

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
```
Comment (lelia) [inline .github/workflows/release.yml:41]: @RutvikS-crest @rquedas FYI that we've rotated the signing keys to use a more descriptive name, as they're now stored at the organizational level. Please accept my commit suggestions so the workflow will use the correct values, thanks!

```suggestion
          gpg_private_key: ${{ secrets.TERRAFORM_GPG_PRIVATE_KEY }}
          passphrase: ${{ secrets.TERRAFORM_GPG_PASSPHRASE }}
```

**Review by lelia — CHANGES_REQUESTED** [review]
@RutvikS-crest any theories on what's causing the [issues](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5520871843?pr=35) with `go generate` and acceptance test jobs?

**Comment by RutvikS-crest** [comment]
> @RutvikS-crest any theories on what's causing the [issues](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5520871843?pr=35) with `go generate` and acceptance test jobs?

@lelia, Tests are failing because they need access to AWS, Azure environments, I have added necessary environment variables in settings. You need to fill those in order to run tests. I have ran tests locally in my account it's running fine.

For now I have removed Workflows related to tests as they are optional to release the provider.

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        name: Run GoReleaser
+        uses: goreleaser/goreleaser-action@f82d6c1c344bcacabba2c841718984797f664a6b # v4.2.0
+        with:
+          version: latest
+          args: release --rm-dist
+        env:
+          GPG_FINGERPRINT: ${{ steps.import_gpg.outputs.fingerprint }}
+          # GitHub sets this automatically
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
```
Comment (lelia) [inline .github/workflows/release.yml:52]: [NIT] missing newline

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,52 @@
+# This GitHub action can publish assets for release when a tag is created.
+# Currently its setup to run on any tag that matches the pattern "v*" (ie. v0.1.0).
+#
+# This uses an action (crazy-max/ghaction-import-gpg) that assumes you set your
+# private key in the `GPG_PRIVATE_KEY` secret and passphrase in the `PASSPHRASE`
+# secret. If you would rather own your own GPG handling, please fork this action
+# or use an alternative one for key handling.
```
Comment (lelia) [inline .github/workflows/release.yml:7]:
```suggestion
# This uses an action (crazy-max/ghaction-import-gpg) that assumes you set your
# private key in the `TERRAFORM_GPG_PRIVATE_KEY` secret and passphrase in the
# `TERRAFORM_GPG_PASSPHRASE` secret. If you would rather own your own GPG handling,
# please fork this action or use an alternative one for key handling.
```

**Comment by lelia** [comment]
> > @RutvikS-crest any theories on what's causing the [issues](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5520871843?pr=35) with `go generate` and acceptance test jobs?
> 
> @lelia, Tests are failing because they need access to AWS, Azure environments, I have added necessary environment variables in settings. You need to fill those in order to run tests. I have ran tests locally in my account it's running fine.
> 
> For now I have removed Workflows related to tests as they are optional to release the provider.

@RutvikS-crest Thanks for the additional context! Could you [file an issue](https://github.com/cisco-open/terraform-provider-appd/issues/new/choose) describing what's currently failing, what needs to be done to fix the tests, and which specific environment variables need to be configured?

@rquedas After we merge this PR, can you work with @RutvikS-crest on resolving the above testing issue once we've successfully released on GitHub / published to the Terraform Registry?

**Review by lelia — CHANGES_REQUESTED** [review]
Thanks for updating the GPG signing key names, @RutvikS-crest! 

Just a couple minor workflow fixes here and I think we'll be good to go.

**Review by lelia — APPROVED** [review]
Latest changes LGTM. @RutvikS-crest can you please file an issue as described in [my comment above](https://github.com/cisco-open/terraform-provider-appd/pull/35#issuecomment-1648681190) before we merge this? Thanks!

**Comment by cginsberg98** [comment]
@lelia and @RutvikS-crest  I am trying to report on the status but seems latest thread was 2 weeks ago, is the action with @RutvikS-crest  to close this out as per message above?

**Pull request merged**